src/h3lib/lib/algos.c

```diff
@@ -634,28 +634,121 @@ int H3_EXPORT(hexRing)(H3Index origin, int k, H3Index* out) {
+ * them and their neighbors to be contained by the geofence(s), and then any
+ * newly found hexagons are used to test again until no new hexagons are found.
  *
  * @param geoPolygon The geofence and holes defining the relevant area
  * @param res The Hexagon resolution (0-15)
  * @param out The slab of zeroed memory to write to. Assumed to be big enough.
  */
 void H3_EXPORT(polyfill)(const GeoPolygon* geoPolygon, int res, H3Index* out) {
+    // TODO: Eliminate this wrapper with the H3 4.0.0 release
+    int failure = _polyfillInternal(geoPolygon, res, out);
+    // LCOV_EXCL_START
+    if (failure) {
+        int numHexagons = H3_EXPORT(maxPolyfillSize)(geoPolygon, res);
+        for (int i = 0; i < numHexagons; i++) out[i] = H3_INVALID_INDEX;
+    }
+    // LCOV_EXCL_STOP
+}
+
+/**
+ * _getEdgeHexagons takes a given geofence ring (either the main geofence or
+ * one of the holes) and traces it with hexagons and updates the search and
+ * found memory blocks. This is used for determining the initial hexagon set
+ * for the polyfill algorithm to execute on.
+ *
+ * @param geofence The geofence (or hole) to be traced
+ * @param numHexagons The maximum number of hexagons possible for the geofence
+ *                    (also the bounds of the search and found arrays)
+ * @param res The hexagon resolution (0-15)
+ * @param numSearchHexes The number of hexagons found so far to be searched
+ * @param search The block of memory containing the hexagons to search from
+ * @param found The block of memory containing the hexagons found from the
+ * search
+ *
+ * @return An error code if the hash function cannot insert a found hexagon
+ *         into the found array.
+ */
+int _getEdgeHexagons(const Geofence* geofence, int numHexagons, int res,
+                     int* numSearchHexes, H3Index* search, H3Index* found) {
+    for (int i = 0; i < geofence->numVerts; i++) {
+        GeoCoord origin = geofence->verts[i];
+        GeoCoord destination = i == geofence->numVerts - 1
+                                   ? geofence->verts[0]
+                                   : geofence->verts[i + 1];
+        const int numHexesEstimate =
+            lineHexEstimate(&origin, &destination, res);
+        for (int j = 0; j < numHexesEstimate; j++) {
+            GeoCoord interpolate;
+            interpolate.lat =
+                (origin.lat * (numHexesEstimate - j) / numHexesEstimate) +
+                (destination.lat * j / numHexesEstimate);
+            interpolate.lon =
+                (origin.lon * (numHexesEstimate - j) / numHexesEstimate) +
+                (destination.lon * j / numHexesEstimate);
+            H3Index pointHex = H3_EXPORT(geoToH3)(&interpolate, res);
+            // A simple hash to store the hexagon, or move to another place if
+            // needed
+            int loc = (int)(pointHex % numHexagons);
+            int loopCount = 0;
+            while (found[loc] != 0) {
```

Comment:
Seems like this kind of hexagon hash set code is being reimlemented in the library, in the future we should see if we can refactoring it.